src/sources/host_metrics/cgroups.rs

```diff
@@ -0,0 +1,271 @@
+                "Failed to load/parse cgroups current memory",
+            ) {
+                result.push(self.gauge("cgroup_memory_current", now, current as f64, tags.clone()));
+            }
+
+            if let Some(stat) = filter_result_sync(
+                cgroup.load_memory_stat(),
+                "Failed to load/parse cgroups memory statistics",
+            ) {
+                result.push(self.gauge("cgroup_memory_anon", now, stat.anon as f64, tags.clone()));
+                result.push(self.gauge("cgroup_memory_file", now, stat.file as f64, tags));
+            }
+        }
+
+        if level < self.cgroups.levels {
+            if let Some(children) =
+                filter_result_sync(cgroup.children(), "Failed to load cgroups children")
+            {
+                for child in children {
+                    if self.cgroups.groups.contains_path(Some(&child.name)) {
+                        result.extend(self.recurse_cgroup(now, child, level + 1));
+                    }
+                }
+            }
+        }
+
+        result
+    }
+}
+
+#[derive(Debug)]
+struct CGroup {
+    root: PathBuf,
+    name: PathBuf,
+}
+
+impl CGroup {
+    fn root<P: AsRef<Path>>(base_group: Option<P>) -> Option<CGroup> {
+        let base_dir = join_path(heim::os::linux::sysfs_root(), "fs/cgroup");
+        let root = join_path(&base_dir, "unified");
+        is_dir(&root)
+            .then(|| root)
+            .or_else(|| is_file(join_path(&base_dir, "cgroup.procs")).then(|| base_dir))
+            .and_then(|root| match base_group {
+                Some(group) => {
+                    let group = group.as_ref();
+                    let root = join_path(root, group);
+                    is_dir(&root).then(|| CGroup {
+                        root,
+                        name: group.into(),
+                    })
+                }
+                None => Some(CGroup {
+                    root,
+                    name: "/".into(),
+                }),
+            })
+    }
+
+    fn is_root(&self) -> bool {
+        self.name == Path::new("/")
+    }
+
+    fn load_cpu(&self) -> io::Result<CpuStat> {
+        let mut data = String::new();
+        File::open(self.make_path("cpu.stat"))?.read_to_string(&mut data)?;
+        data.parse()
+            .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))
+    }
+
+    fn make_path(&self, filename: impl AsRef<Path>) -> PathBuf {
+        join_path(&self.root, filename)
+    }
+
+    fn load_memory_current(&self) -> io::Result<u64> {
+        let mut current = String::new();
+        File::open(self.make_path("memory.current"))?.read_to_string(&mut current)?;
+        current.trim().parse().map_err(map_parse_error)
+    }
+
+    fn load_memory_stat(&self) -> io::Result<MemoryStat> {
+        let mut data = String::new();
+        File::open(self.make_path("memory.stat"))?.read_to_string(&mut data)?;
+        data.parse().map_err(map_parse_error)
+    }
```

Comment:
To my comment [here](https://github.com/timberio/vector/pull/8806/files#r694121728) these ought to be async based on who their eventual caller is. Eliding the temporary string allocation by parsing from the `Read` directly would also be cool. IIRC this format is intended for such things.

Reply:
Making it all async gets complicated because the cgroups structure is recursive. Recursive async functions are not allowed. So, we would have to drop support for reporting anything beyond the base requested group, or handle the hierarchy by implementing the recursion stack manually. I have implemented reusing the buffer in aadd968ee.